nexus/src/app/sagas/instance_migrate.rs

```diff
@@ -21,36 +27,115 @@ use uuid::Uuid;
+    // CRDB by asking the instance's current sled to paste them into its runtime
+    // state. Sled agent provides the synchronization here: while this operation
+    // is idempotent for any single transition between IDs, sled agent ensures
+    // that if multiple concurrent sagas try to set migration IDs at the same
+    // Propolis generation, then only one will win and get to proceed through
+    // the saga.
+    SET_MIGRATION_IDS -> "set_migration_ids" {
+        + sim_set_migration_ids
     }
-    INSTANCE_MIGRATE -> "instance_migrate" {
-        // TODO robustness: This needs an undo action
-        + sim_instance_migrate
+
+    // The instance state on the destination looks like the instance state on
+    // the source, except that it bears all of the destination's "location"
+    // information--its Propolis ID, sled ID, and Propolis IP--with the same
+    // Propolis generation number as the source set in the previous step.
+    // Consider an example:
+    //
+    // - Before the saga begins, the instance has Propolis generation 10 and no
+    //   migration IDs.
+    // - The previous step sets Propolis generation 11 and sets the instance's
+    //   migration ID.
+    // - This step synthesizes a record with Propolis generation 11 and the same
+    //   migration IDs, but with the Propolis identifiers set to the
+    //   destination's IDs.
+    //
+    // When the migration resolves one way or the other, one or both of the
+    // participating sleds will update CRDB with Propolis generation 12 and the
+    // Propolis ID set to whichever Propolis ended up running the instance.
+    //
+    // This step must be infallible and does not have an undo action for reasons
+    // described below.
+    CREATE_DESTINATION_STATE -> "dst_runtime_state" {
+        + sim_create_destination_state
     }
-    V2P_ENSURE -> "v2p_ensure" {
-        // TODO robustness: This needs an undo action
-        + sim_v2p_ensure
+
+    // The next three steps are organized so that if the saga unwinds after the
+    // destination Propolis is created, the "set migration IDs" step is undone
+    // before the "ensure destination Propolis" step. Consider again the example
+    // above, but this time, let the saga unwind after the destination Propolis
+    // starts. If sled agent changes an instance's Propolis generation when a
+    // Propolis is destroyed, and the undo steps are specified in the
+    // traditional way, the following can occur:
+    //
+    // - After the steps above, the instance is registered with the source and
+    //   target sleds, both at Propolis generation 11.
+    // - Undoing the "ensure destination Propolis" step sets the Propolis
+    //   generation to 12, but with the location information from the target!
+    // - When the "set migration IDs" step is undone, the source will publish
+    //   Propolis generation 12, but this will be dropped because the generation
+    //   was already incremented.
```

Comment:
The thing we're trying to do here would be to *clear* the migration ID, right?

If we know this happens on the undo path, couldn't we just bump the generation number before sending the request?

Reply:
2a1cad71 should clean this up as well.